sample_size/metrics.py

```diff
@@ -1,19 +1,32 @@
 from abc import ABCMeta
 from abc import abstractmethod
 from typing import Union
+from math import sqrt
 
 from statsmodels.stats.power import NormalIndPower
 from statsmodels.stats.power import TTestIndPower
 
 
 class BaseMetric:
-
     __metaclass__ = ABCMeta
     mde: float
 
     def __init__(self, mde: float):
         self.mde = mde
 
+    def single_sample_size(self, alpha: float, power: float) -> int:
```

Comment:
this is duplicate with `SampleSizeCalculator._get_single_sample_size`. We should keep sample size calculation in the SampleSizeCalculator

Reply:
@dfu-godaddy I think this is a direct response to [this comment I made](https://github.com/godaddy/sample-size/pull/24#pullrequestreview-959656809). However, I think there might have been some miscommunication on my part. 

What I was suggesting was something more like

```python3
class BaseMetric:
    def single_sample_size(self) -> float:
        raise NotImplementedError

class Numeric(BaseMetric):
    def single_sample_size(self):
        effect_size = self.mde / float(sqrt(self.variance))

         return int(
             stats.ttest.solve_power(
                 effect_size=effect_size,
                 alpha=alpha,
                 power=power,
                 ratio=1,
                 alternative="two-sided",
             )
         )
```
```python3
class SampleSizeCalculator:
    def get_sample_size(self) -> float:
        return self.get_multiple_sample_size()

    def get_multiple_sample_size(self) -> int:
        if len(self.metrics) < 2:
            return self.metrics[0].single_sample_size()
        lower = max([metric.single_sample_size(self.alpha, self.power) for metric in self.metrics])
        upper = max([metric.single_sample_size(self.alpha / self.m, self.power) for metric in self.metrics])

        return self._find_sample_size(lower, upper)
```

What @dfu-godaddy is saying about duplication is true, and we should avoid it. In terms of keeping power-calculation encapsulated in one class, that could also work
```python3
class BaseMetric:
    # no single_sample_size
       

class Numeric(BaseMetric):
    # no single_sample_size
```

```python3
class SampleSizeCalculator:
    def get_sample_size(self) -> int:
        return self.get_multiple_sample_size()

    def _single_sample_size(self, metric) -> int:
       if type(metric) == Numeric:
           return stats.ttest.solve_power(
              ...
           )
       elif type(metric) == Ratio:
           return ...
 

    def get_multiple_sample_size(self) -> int:
        if len(self.metrics) < 2:
            return self.metrics[0].single_sample_size()

        lower = max([metric.single_sample_size(self.alpha, self.power) for metric in self.metrics])
        upper = max([metric.single_sample_size(self.alpha / self.m, self.power) for metric in self.metrics])

        return self._find_sample_size(lower, upper)
```

I can live with the the second implementation, but generally if you're using a type-based switch-case statement, it's better to make the functionality of that switch-case _a property of the type_

Reply:
> @dfu-godaddy I think this is a direct response to [this comment I made](https://github.com/godaddy/sample-size/pull/24#pullrequestreview-959656809). However, I think there might have been some miscommunication on my part.
> 
> What I was suggesting was something more like
> 
> ```python
> class BaseMetric:
>     def single_sample_size(self) -> float:
>         raise NotImplementedError
> 
> class Numeric(BaseMetric):
>     def single_sample_size(self):
>         effect_size = self.mde / float(sqrt(self.variance))
> 
>          return int(
>              stats.ttest.solve_power(
>                  effect_size=effect_size,
>                  alpha=alpha,
>                  power=power,
>                  ratio=1,
>                  alternative="two-sided",
>              )
>          )
> ```
> 
>  ```python
> class SampleSizeCalculator:
>     def get_sample_size(self) -> float:
>         return self.get_multiple_sample_size()
> 
>     def get_multiple_sample_size(self) -> int:
>         if len(self.metrics) < 2:
>             return self.metrics[0].single_sample_size()
>         lower = max([metric.single_sample_size(self.alpha, self.power) for metric in self.metrics])
>         upper = max([metric.single_sample_size(self.alpha / self.m, self.power) for metric in self.metrics])
> 
>         return self._find_sample_size(lower, upper)
> ```
> 
> What @dfu-godaddy is saying about duplication is true, and we should avoid it. In terms of keeping power-calculation encapsulated in one class, that could also work
> 
> ```python
> class BaseMetric:
>     # no single_sample_size
>        
> 
> class Numeric(BaseMetric):
>     # no single_sample_size
> ```
> 
>  ```python
> class SampleSizeCalculator:
>     def get_sample_size(self) -> int:
>         return self.get_multiple_sample_size()
> 
>     def _single_sample_size(self, metric) -> int:
>        if type(metric) == Numeric:
>            return stats.ttest.solve_power(
>               ...
>            )
>        elif type(metric) == Ratio:
>            return ...
>  
> 
>     def get_multiple_sample_size(self) -> int:
>         if len(self.metrics) < 2:
>             return self.metrics[0].single_sample_size()
> 
>         lower = max([metric.single_sample_size(self.alpha, self.power) for metric in self.metrics])
>         upper = max([metric.single_sample_size(self.alpha / self.m, self.power) for metric in self.metrics])
> 
>         return self._find_sample_size(lower, upper)
> ```
> 
> I can live with the the second implementation, but generally if you're using a type-based switch-case statement, it's better to make the functionality of that switch-case _a property of the type_

It seems that my last changes on this function was missed from the commit. I did delete get_single_sample_size from `sample_size_calculator.py` and make `get_sample_size` as follows:
```
def get_sample_size(self) -> float:
        if len(self.metrics) == 1 and self.variants == DEFAULT_VARIANTS:
            sample_size = self.metrics[0].get_single_sample_size(self.alpha, self.power)
```

Also why can't we make it a BaseMetric method instead of a property needed to be implemented by each metric class?

Reply:
> Why can't we make it a BaseMetric method 

I suppose it comes down to personal preference. But, in general, I'm not sure the construct `default_power_analysis_instance` completely makes sense to me. Specifically its unclear how we would choose a sensible default / what situation would necessitate inferring a metric type. Defining the implementation in the base class means you have reason about inheritance and remember what's been defined and what's being overridden. In a sense, its a "composition-vs-inheritance" issue. In context the only reason to define some inherited power-analysis instance is to avoid redundancy between the binomial and numeric cases which both use a z-test, but the calculation of effect size and variance is different between those cases. The way I see it, for the 3 metrics we have, there are 3 distinct power calculations.

But at the end of the day, this is not a blocker.